internal/rbd/snapshot.go

```diff
@@ -0,0 +1,122 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbd
+
+import (
+	"context"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+func createRBDClone(ctx context.Context, parentVol, cloneRbdVol *rbdVolume, snap *rbdSnapshot, cr *util.Credentials) error {
+	// create snapshot
+	err := createSnapshot(ctx, parentVol, snap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s: %v"), snap, err)
+		return err
+	}
+
+	// create clone image and delete snapshot
+	err = cloneRbdImageFromSnapshot(ctx, cloneRbdVol, snap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone rbd image %s from snapshot %s: %v"), cloneRbdVol.RbdImageName, snap.RbdSnapName, err)
+		err = errors.Errorf("failed to clone rbd image %s from snapshot %s: %v", cloneRbdVol.RbdImageName, snap.RbdSnapName, err)
+	}
+	errSnap := deleteSnapshot(ctx, parentVol, snap)
+	if errSnap != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errSnap)
+		delErr := deleteImage(ctx, cloneRbdVol, cr)
+		if delErr != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"), cloneRbdVol, delErr)
+		}
+		return err
+	}
+
+	err = cloneRbdVol.getImageInfo()
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to get rbd image: %s details with error: %v"), cloneRbdVol, err)
+		delErr := deleteImage(ctx, cloneRbdVol, cr)
+		if delErr != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"), cloneRbdVol, delErr)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func cleanUpSnapshot(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, rbdVol *rbdVolume, cr *util.Credentials) error {
+	_, err := getSnapInfo(parentVol, rbdSnap)
+	if err != nil {
+		if _, ok := err.(ErrSnapNotFound); !ok {
+			return err
+		}
+	} else {
+		err = deleteSnapshot(ctx, parentVol, rbdSnap)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), err)
+			return err
+		}
+	}
+
+	err = rbdVol.getImageInfo()
+	if err != nil {
+		if _, ok := err.(ErrImageNotFound); !ok {
+			return err
+		}
+	} else {
+		err = deleteImage(ctx, rbdVol, cr)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s/%s with error: %v"), rbdVol.Pool, rbdVol.VolName, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func generateVolFromSnap(rbdSnap *rbdSnapshot) *rbdVolume {
+	vol := new(rbdVolume)
+	vol.ClusterID = rbdSnap.ClusterID
+	vol.VolID = rbdSnap.SnapID
+	vol.Monitors = rbdSnap.Monitors
+	vol.Pool = rbdSnap.Pool
+	vol.RbdImageName = rbdSnap.RbdSnapName
+	return vol
+}
+
+func undoSnapshotCloning(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cloneVol *rbdVolume, cr *util.Credentials, deleteVol bool) error {
+	err := deleteSnapshot(ctx, parentVol, rbdSnap)
```

Comment:
This is the snapshot of the cloned image, that would be used for future CSI-cloning, right? Attempting to get my bearing straight on this one :)